### PR TITLE
cts, mts, cjs, mjs, etc extension support

### PR DESCRIPTION
**Pull request by weswigham**
This is my first followup to 
https://github.com/microsoft/TypeScript/pull/44501 which officially adds support for loading `.mjs`, `.cjs`, `.mts`, `.cts`, `.d.mts`, and `.d.cts` files. In non-`node` `moduleResolution` modes, these are largely the same as `.js`, `.ts`, and `.d.ts` - the only difference being the output path mappings - a `.mts` or `.mjs` input always makes a `.mjs` and `.d.mts` output, while a `.cjs` or `.cts` input always makes a `.cjs` and `.d.cts` output. In `node`ish module resolution modes, `c`-prefixed files imply a commonjs emit format, while `m`-prefixed files imply an esm format. As we've discussed in design meetings, to avoid comitting to a direction, we also forbid ambiguous generic arrow functions and angle bracket type cast syntax in `.cts` and `.mts` files, so we can reserve the right to make these files parse as jsx in the future.

The watch/build mode baseline tests currently don't show any changes with edits, due to the bug fixed in https://github.com/microsoft/TypeScript/pull/44935 - that'll need to get pulled in, and, once it is, the baseline updated.

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -5975,6 +5975,14 @@
         "category": "Error",
         "code": 7057
     },
+    "This syntax is reserved in files with this extension. Use an `as` expression instead.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: This is slightly better, but only worthwhile if it's pretty easy:

```suggestion
    "This syntax is reserved in files with the extension {0}. Use an `as` expression instead.": {
```

(and for the other message too)

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -106,26 +106,23 @@ namespace ts {
         return (options.sourceMap && !options.inlineSourceMap) ? jsFilePath + ".map" : undefined;
     }
 
-    // JavaScript files are always LanguageVariant.JSX, as JSX syntax is allowed in .js files also.
-    // So for JavaScript files, '.jsx' is only emitted if the input was '.jsx', and JsxEmit.Preserve.
-    // For TypeScript, the only time to emit with a '.jsx' extension, is on JSX input, and JsxEmit.Preserve
     /* @internal */
-    export function getOutputExtension(sourceFile: SourceFile, options: CompilerOptions): Extension {
-        if (isJsonSourceFile(sourceFile)) {
+    export function getOutputExtension(fileName: string, options: CompilerOptions): Extension {
```
Comment (sandersn) [inline src/compiler/emitter.ts:110]: why is this function separate from getDeclarationEmitExtensionForPath?

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -106,26 +106,23 @@ namespace ts {
...
                 return Extension.Jsx;
             }
         }
+        if (fileExtensionIsOneOf(fileName, [Extension.Mts, Extension.Mjs])) {
+            return Extension.Mjs;
+        }
+        if (fileExtensionIsOneOf(fileName, [Extension.Cts, Extension.Cjs])) {
+            return Extension.Cjs;
+        }
         return Extension.Js;
```
Comment (sandersn) [inline src/compiler/emitter.ts]: minor, but a single-expression rewrite to look like getDeclarationEmitExtensionForPath would be easier to read

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -6702,56 +6710,61 @@ namespace ts {
...
         ];
 
-        return deduplicate<string>(extensions, equateStringsCaseSensitive, compareStringsCaseSensitive);
+        return extensions;
     }
 
-    export function getSuppoertedExtensionsWithJsonIfResolveJsonModule(options: CompilerOptions | undefined, supportedExtensions: readonly string[]): readonly string[] {
+    export function getSuppoertedExtensionsWithJsonIfResolveJsonModule(options: CompilerOptions | undefined, supportedExtensions: readonly Extension[][]): readonly Extension[][];
+    export function getSuppoertedExtensionsWithJsonIfResolveJsonModule(options: CompilerOptions | undefined, supportedExtensions: readonly string[][]): readonly string[][];
+    export function getSuppoertedExtensionsWithJsonIfResolveJsonModule(options: CompilerOptions | undefined, supportedExtensions: readonly string[][]): readonly string[][] {
```
Comment (sandersn) [inline src/compiler/utilities.ts]: typo: Supported

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -6702,56 +6710,61 @@ namespace ts {
...
-    export const supportedJSExtensions: readonly Extension[] = [Extension.Js, Extension.Jsx];
-    export const supportedJSAndJsonExtensions: readonly Extension[] = [Extension.Js, Extension.Jsx, Extension.Json];
-    const allSupportedExtensions: readonly Extension[] = [...supportedTSExtensions, ...supportedJSExtensions];
-    const allSupportedExtensionsWithJson: readonly Extension[] = [...supportedTSExtensions, ...supportedJSExtensions, Extension.Json];
-
-    export function getSupportedExtensions(options?: CompilerOptions): readonly Extension[];
-    export function getSupportedExtensions(options?: CompilerOptions, extraFileExtensions?: readonly FileExtensionInfo[]): readonly string[];
-    export function getSupportedExtensions(options?: CompilerOptions, extraFileExtensions?: readonly FileExtensionInfo[]): readonly string[] {
+    const supportedTSExtensionsForExtractExtension: readonly Extension[] = [Extension.Dts, Extension.Dcts, Extension.Dmts, Extension.Cts, Extension.Mts, Extension.Ts, Extension.Tsx, Extension.Cts, Extension.Mts];
+    export const supportedJSExtensions: readonly Extension[][] = [[Extension.Js, Extension.Jsx], [Extension.Mjs], [Extension.Cjs]];
```
Comment (sandersn) [inline src/compiler/utilities.ts:6760]: I can't find where the non-flat version is used.

File: src/server/watchType.ts [hunk 6]
```diff
@@ -8,7 +8,6 @@ namespace ts {
         MissingSourceMapFile: "Missing source map file",
         NoopConfigFileForInferredRoot: "Noop Config file for the inferred project root",
         MissingGeneratedFile: "Missing generated file",
-        PackageJsonFile: "package.json file for import suggestions",
```
Comment (sandersn) [inline src/server/watchType.ts:11]: What does this do? Why does it go away?

File: src/services/stringCompletions.ts [hunk 7]
```diff
@@ -344,10 +344,10 @@ namespace ts.Completions.StringCompletions {
         }
     }
 
-    function getSupportedExtensionsForModuleResolution(compilerOptions: CompilerOptions): readonly Extension[] {
+    function getSupportedExtensionsForModuleResolution(compilerOptions: CompilerOptions): readonly Extension[][] {
```
Comment (sandersn) [inline src/services/stringCompletions.ts:347]: this is only called one place, and is immediately flattened. Why not flatten here?

Edit: All callers of getSuppoertedExtensionsWithJsonIfResolveJsonModule and getSupportedExtensions also immediately flatten. Maybe they should flatten in there. Where is the non-flattened version of these lists used?

File: src/testRunner/unittests/moduleResolution.ts [hunk 8]
```diff
@@ -81,9 +81,11 @@ namespace ts {
     }
 
     describe("unittests:: moduleResolution:: Node module resolution - relative paths", () => {
-
+        // node module resolution does _not_ implicitly append these extensions to an extensionless path (though will still attempt to load them if explicitly)
+        const nonImplicitExtensions = [Extension.Mts, Extension.Dmts, Extension.Mjs, Extension.Cts, Extension.Dcts, Extension.Cjs];
+        const autoExtensions = filter(supportedTSExtensionsFlat, e => nonImplicitExtensions.indexOf(e) === -1);
```
Comment (sandersn) [inline src/testRunner/unittests/moduleResolution.ts:87]: I wish we could use Array.includes here, it's much easier to read.

File: src/compiler/commandLineParser.ts [hunk 9]
```diff
@@ -3439,16 +3439,24 @@ namespace ts {
...
-            const higherPriorityExtension = extensions[i];
-            const higherPriorityPath = keyMapper(changeExtension(file, higherPriorityExtension));
+    function hasFileWithHigherPriorityExtension(file: string, literalFiles: ESMap<string, string>, wildcardFiles: ESMap<string, string>, extensions: readonly string[][], keyMapper: (value: string) => string) {
+        const extensionGroup = forEach(extensions, group => fileExtensionIsOneOf(file, group) ? group : undefined);
+        if (!extensionGroup) {
+            return false;
+        }
+        for (const ext of extensionGroup) {
+            if (fileExtensionIs(file, ext)) {
+                break;
```
Comment (sandersn) [inline src/compiler/commandLineParser.ts]: These are equivalent, right?

```suggestion
                return false;
```

File: src/compiler/commandLineParser.ts [hunk 10]
```diff
@@ -3439,16 +3439,24 @@ namespace ts {
...
+            }
+            const higherPriorityPath = keyMapper(changeExtension(file, ext));
```
Comment (sandersn) [inline src/compiler/commandLineParser.ts:3469]: What's an example of a keyMapper? What do they usually do?

File: src/compiler/commandLineParser.ts [hunk 11]
```diff
@@ -3439,16 +3439,24 @@ namespace ts {
...
-     * @param extensionPriority The priority of the extension.
-     * @param context The expansion context.
      */
-    function hasFileWithHigherPriorityExtension(file: string, literalFiles: ESMap<string, string>, wildcardFiles: ESMap<string, string>, extensions: readonly string[], keyMapper: (value: string) => string) {
-        const extensionPriority = getExtensionPriority(file, extensions);
-        const adjustedExtensionPriority = adjustExtensionPriority(extensionPriority, extensions);
-        for (let i = ExtensionPriority.Highest; i < adjustedExtensionPriority; i++) {
-            const higherPriorityExtension = extensions[i];
-            const higherPriorityPath = keyMapper(changeExtension(file, higherPriorityExtension));
+    function hasFileWithHigherPriorityExtension(file: string, literalFiles: ESMap<string, string>, wildcardFiles: ESMap<string, string>, extensions: readonly string[][], keyMapper: (value: string) => string) {
```
Comment (sandersn) [inline src/compiler/commandLineParser.ts:3460]: is this and removeWildcardFilesWithLowerPriorityExtension the only place where the new `string[][]` structure of `extensions` is used?

File: src/compiler/commandLineParser.ts [hunk 12]
```diff
@@ -3461,15 +3469,18 @@ namespace ts {
...
-            const lowerPriorityPath = keyMapper(changeExtension(file, lowerPriorityExtension));
+    function removeWildcardFilesWithLowerPriorityExtension(file: string, wildcardFiles: ESMap<string, string>, extensions: readonly string[][], keyMapper: (value: string) => string) {
+        const extensionGroup = forEach(extensions, group => fileExtensionIsOneOf(file, group) ? group : undefined);
+        if (!extensionGroup) {
+            return;
+        }
+        for (let i = extensionGroup.length - 1; i >= 0; i--) {
+            const ext = extensionGroup[i];
+            if (fileExtensionIs(file, ext)) {
+                break;
```
Comment (sandersn) [inline src/compiler/commandLineParser.ts]:
```suggestion
                return;
```

File: src/compiler/utilities.ts [hunk 13]
```diff
@@ -6702,56 +6710,61 @@ namespace ts {
     /**
      *  List of supported extensions in order of file resolution precedence.
```
Comment (sandersn) [inline src/compiler/utilities.ts]: can you update this comment to explain what the multiple lists mean?

**Review by sandersn — COMMENTED** [review]
Initial comments -- mostly questions about the new `Extension[][]` structure for supported extensions.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
D'oh.

File: src/compiler/emitter.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/emitter.ts:110]: JS output extension vs declaration file output extension.

File: src/server/watchType.ts (hunk 6 above)
Comment (weswigham) [reply src/server/watchType.ts:11]: Duplicated by the `PackageJson` type which is now used by the base watcher.

File: src/services/stringCompletions.ts (hunk 7 above)
Comment (weswigham) [reply src/services/stringCompletions.ts:347]: I mean, it's one function call used in one place - I don't think it matters much, except for the sake of future modification deferring flatting for a long as possible is preferable.

File: src/compiler/commandLineParser.ts (hunk 10 above)
Comment (weswigham) [reply src/compiler/commandLineParser.ts:3469]: Canonicalizes the filename.

File: src/compiler/commandLineParser.ts (hunk 11 above)
Comment (weswigham) [reply src/compiler/commandLineParser.ts:3460]: Pretty much! It's the raison d'être for both the old priority system, and the new ordered groups.

**Comment by orta** [comment]
Might be worth making this a separate draft PR to TypeScript so we can get `pack this` working?

**Comment by weswigham** [comment]
https://github.com/microsoft/TypeScript/pull/45774 should track this on the TS repo and be a place where you can run bot commands (versus usefully review code)

File: src/compiler/utilities.ts [hunk 14]
```diff
@@ -6742,58 +6750,63 @@ namespace ts {
     }
 
     /**
-     *  List of supported extensions in order of file resolution precedence.
+     *  Groups of supported extensions in order of file resolution precedence. (eg, TS > TSX > DTS and seperately, CTS > DCTS)
```
Comment (sandersn) [inline src/compiler/utilities.ts:6753]:
```suggestion
     *  Groups of supported extensions in order of file resolution precedence. (eg, TS > TSX > DTS and separately, CTS > DCTS)
```

**Review by sandersn — APPROVED** [review]

**Review by andrewbranch — APPROVED** [review]

**Comment by weswigham** [comment]
Merged in #45884

**Pull request closed without merging**